ipc_types/ipc_subnet_actor.go

```diff
@@ -0,0 +1,190 @@
+	checkpoint, err := utils.GetOutOfHamt[Checkpoint](st.Stake, s, abi.AddrKey(id))
+	return checkpoint, err
+}
+
+func (st *IPCSubnetActorState) GetWindowCheck(s adt.Store, id address.Address) (*Votes, error) {
+	votes, err := utils.GetOutOfHamt[Votes](st.Stake, s, abi.AddrKey(id))
+	return votes, err
+}
+
+func (st *IPCSubnetActorState) GetCrossMsgs(s adt.Store, c cid.Cid) (*CrossMsgs, error) {
+	crossMsgs, err := utils.GetOutOfHamt[CrossMsgs](st.Stake, s, abi.CidKey(c))
+	return crossMsgs, err
+}
+
+func (st *IPCSubnetActorState) HasMajorityVote(s adt.Store, v Votes) (bool, error) {
+	sum := big.Zero()
+	for _, m := range v.Validators {
+		stake, err := st.GetStake(s, m)
+		if err != nil {
+			return false, err
+		}
+		sum = big.Sum(sum, *stake)
+	}
+	fsum := new(mbig.Rat).SetInt(sum.Int)
+	fTotal := new(mbig.Rat).SetInt(st.TotalStake.Int)
+	div := new(mbig.Rat).SetFrac(fsum.Num(), fTotal.Num())
+	threshold := utils.MajorityThreshold()
+	return div.Cmp(&threshold) >= 0, nil
+
+}
+
+func (s *IPCSubnetActorState) IsValidator(addr *address.Address) bool {
+	for _, x := range s.ValidatorSet {
+		if x.Addr == *addr {
+			return true
+		}
+	}
+	return false
+}
+
+func getMsgMeta(msgMetas *adt.Map, c cid.Cid) (*CrossMsgs, bool, error) {
+	var out CrossMsgs
+	found, err := msgMetas.Get(abi.CidKey(c), &out)
+	if err != nil {
+		return nil, false, xerrors.Errorf("failed to get crossMsgMeta from registry with cid %v: %w", c, err)
+	}
+	if !found {
+		return nil, false, nil
+	}
+	return &out, true, nil
+}
+
+type ConsensusType int64
+
+const (
+	Delegated ConsensusType = iota
+	PoW
+	Tendermint
+	Mir
+	FilecoinEC
+	Dummy
+)
+
+type Status int64
+
+const (
+	Instantiated Status = iota
+	Active
+	Inactive
+	Terminating
+	Killed
+)
+
+type ChainEpoch int64
+
+type Validator struct {
+	Addr    address.Address
+	NetAddr string
+}
+
+type ConstructParams struct {
+	Parent            SubnetID
+	Name              string
+	IpcGatewayAddr    uint64
+	Consensus         ConsensusType
+	MinValidatorStake abi.TokenAmount
+	MinValidators     uint64
+	FinalityThreshold ChainEpoch
+	CheckPeriod       ChainEpoch
+	Genesis           []byte
+}
+
+type JoinParams struct {
+	ValidatorNetAddr string
+}
+
+type Votes struct {
+	Validators []address.Address
+}
+
+type Checkpoint struct {
+	Data CheckData
+	Sig  []byte
+}
+
+func (c *Checkpoint) CrossMsgMeta(from, to *SubnetID) (*CrossMsgMeta, bool) {
+	for i, m := range c.Data.CrossMsgs {
+		if *from == m.From && *to == m.To {
+			return &c.Data.CrossMsgs[i], true
+		}
+	}
+	return nil, false
+}
+
+func (s *Checkpoint) CrossMsgMetaIndex(from, to *SubnetID) (int, bool) {
+	for i, m := range s.Data.CrossMsgs {
+		if *from == m.From && *to == m.To {
+			return i, true
+		}
+	}
+	return 0, false
+}
+
+func CheckpointEpoch(epoch, period ChainEpoch) ChainEpoch {
+	return (epoch / period) * period
+}
+
+func WindowEpoch(epoch, period ChainEpoch) ChainEpoch {
+	ind := epoch / period
+	return period * (ind + 1)
+}
+
+type CheckData struct {
```

Comment:
We are missing some methods for some of the checkpoint-related types (mainly the important is `Cid()`). Let's maybe add `Cid()` for the ones that implement it and we can add the pending ones in the future PRs once we kick-off the integration.

Reply:
I did not implement those rust methods as they were only public within the crate. What is an example of (even if updated) implementation of a cid() method for a struct like one of these?